### PR TITLE
Add terraform import script

### PR DESCRIPTION
**Pull request by ichung08**
## Description

Add an import script that will enable users to migrate all of their resources to be managed by terraform. The script will read from existing resources (by listing all orgs, deployments, clusters, etc) and then correctly import them into terraform.

How the script works:
1. Parses all the resources to be imported (which are passed in as args)
2. For each type of resource (eg. Workspace, Cluster, Deployment etc), it gets all the entity ids
3. For each entity id, it generates a Terraform import block in the form of:
```
import {
	id = "%v"
	to = astro_resource.resource_%v
}
```
4. After generating all the import blocks, it concatenates all the outputs and writes it to a file, `import.tf`
5. Once the import blocks are written, it then generates the Terraform HCL configuration using `terraform plan -generate-config-out=generated.tf`, writing all Terraform resource configs to a file, `generated.tf`
6. Lastly, the script auto approves and runs `terraform apply` on the new resources, successfully importing all resources!

Example `import.tf` file:
```
terraform {
	required_providers {
		astro = {
			source = "registry.terraform.io/astronomer/astro"
		}
	}
}

provider "astro" {
	organization_id = "clx42kkcm01fo01o06agtmshg"
	host = "https://api.astronomer-dev.io"
}

import {
	id = "clzlybvvw00kv01j6o0fq77ep"
	to = astro_workspace.workspace_clzlybvvw00kv01j6o0fq77ep
}

import {
	id = "clx42trhg01go01o07e0dozht"
	to = astro_cluster.cluster_clx42trhg01go01o07e0dozht
}

import {
	id = "clx42ugkb01gq01od03k7c49a"
	to = astro_cluster.cluster_clx42ugkb01gq01od03k7c49a
}

import {
	id = "clzawipbm00bm01qw98vzzoca"
	to = astro_user_roles.user_clzawipbm00bm01qw98vzzoca
}
```

Example `generated.tf` file:
```
resource "astro_workspace" "workspace_clzlm6cev00ki01lyvqyyoaal" {
  cicd_enforced_default = false
  description           = "Created by Terraform Acceptance Test - should self-cleanup but can delete manually if needed after 2 hours."
  name                  = "TFAcceptanceTest_QPROZQXNVZ_workspace"
}

# __generated__ by Terraform from "clx42trhg01go01o07e0dozht"
resource "astro_cluster" "cluster_clx42trhg01go01o07e0dozht" {
  cloud_provider        = "AWS"
  name                  = "Terraform_AWS_Cluster_DND"
  pod_subnet_range      = null
  region                = "us-east-1"
  service_peering_range = null
  service_subnet_range  = null
  timeouts              = null
  type                  = "DEDICATED"
  vpc_subnet_range      = "172.20.0.0/20"
  workspace_ids         = []
}

# __generated__ by Terraform from "cl7qqe4tf264442d28fttoe7g8"
resource "astro_user_roles" "user_cl7qqe4tf264442d28fttoe7g8" {
  deployment_roles  = null
  organization_role = "ORGANIZATION_OWNER"
  user_id           = "cl7qqe4tf264442d28fttoe7g8"
  workspace_roles   = null
}
```

## 🎟 Issue(s)

#115 

[Migration Doc](https://www.notion.so/astronomerio/Terraform-Data-Migration-ec77d37fd2f040c9916b07cbdee37fc4)

## 🧪 Functional Testing

**Workspace**
Screenshot 2024-08-14 at 7 17 33 PM

**Cluster**
Screenshot 2024-08-14 at 7 22 08 PM

**Hybrid Cluster Workspace Authorization**
No hybrid clusters to test with

**Deployment**

**Team**
Screenshot 2024-08-14 at 7 33 51 PM

**Team Roles**
Screenshot 2024-08-14 at 7 34 36 PM

**API Token**
Screenshot 2024-08-14 at 7 36 25 PM

**User Roles**
Screenshot 2024-08-14 at 7 38 19 PM

**Importing Multiple Resources**
Screenshot 2024-08-14 at 7 43 17 PM

## 📋 Checklist

- [ ] Added/updated applicable tests
- [ ] Added/updated examples in the `examples/` directory
- [ ] Updated any related [documentation](https://github.com/astronomer/docs/)

### CODE REVIEW

File: import/import.go [hunk 1]
```diff
@@ -0,0 +1,594 @@
...
+	}
+
+	// terraform apply to import the resources
+	cmd = exec.Command("terraform", "apply", "-auto-approve")
+	cmd.Stdout = os.Stdout
+	cmd.Stderr = os.Stderr
+
+	if err := cmd.Run(); err != nil {
+		return err
+	}
```
Comment (ichung08) [inline import/import.go:176]: currently, script is set to automatically run `terraform apply` on the resources - do we want to keep it this way? or give users more control and let them decide what they want to import?

**Comment by ichung08** [comment]
Any go logging library recommendations? The logging works, but want to improve the look of the logs

File: import/import.go [hunk 2]
```diff
@@ -0,0 +1,534 @@
...
+	Resource string
+	Error    error
+}
+
+func main() {
+	log.SetFlags(0)
+	log.Println("Terraform Import Script Starting")
+
+	// collect all arguments from the user, indicating all the resources that need to be imported
+	resourcesPtr := flag.String("resources", "", "Comma separated list of resources to import")
```
Comment (vandyliu) [inline import/import.go:33]: we should have enums and validation on these

Comment (vandyliu) [reply import/import.go:33]: ie if someone puts in something incorrect like `wokrspace`, we should fail and tell them `wokrspace` is invalid and these are the valid ones

File: import/import.go [hunk 3]
```diff
@@ -0,0 +1,534 @@
...
+	}
+
+	// set the organization ID
+	organizationId := *organizationIdPtr
+	if organizationId == "" {
+		log.Fatalf("Organization ID not provided")
+	}
+
+	log.Printf("Using API host: %s", host)
+	log.Printf("Using API token: %s", token)
```
Comment (vandyliu) [inline import/import.go:72]: prob shouldn't print this out

File: import/import.go (hunk 1 above)
Comment (vandyliu) [reply import/import.go:199]: we should not apply it
just give them the file

File: import/import.go [hunk 4]
```diff
@@ -0,0 +1,534 @@
...
+	} else {
+		host = *hostPtr
+	}
+
+	// set the organization ID
+	organizationId := *organizationIdPtr
+	if organizationId == "" {
+		log.Fatalf("Organization ID not provided")
+	}
+
```
Comment (vandyliu) [inline import/import.go:70]: we should make sure they have terraform installed early by checking `terraform -version` here and erroring with a good err message if now

File: import/import.go [hunk 5]
```diff
@@ -0,0 +1,534 @@
...
+func main() {
+	log.SetFlags(0)
+	log.Println("Terraform Import Script Starting")
+
+	// collect all arguments from the user, indicating all the resources that need to be imported
+	resourcesPtr := flag.String("resources", "", "Comma separated list of resources to import")
+	tokenPtr := flag.String("token", "", "API token to authenticate with the platform")
+	hostPtr := flag.String("host", "https://api.astronomer.io", "API host to connect to")
+	organizationIdPtr := flag.String("organizationId", "", "Organization ID to import resources into")
+
```
Comment (vandyliu) [inline import/import.go:37]: we should also have a help flag

File: import/import.go
Comment (vandyliu) [inline import/import.go:1]: we should also have GH actions to build and release this (simple one where we just upload all the binaries to github) version number can just be the date

File: .github/workflows/testacc.yml [hunk 6]
```diff
@@ -111,6 +111,7 @@ jobs:
         terraform:
           - latest
         provider_version:
+          - v0.3.1
```
Comment (vandyliu) [inline .github/workflows/testacc.yml:114]: we should only keep the last 3 major versions so we can prob remove v0.3.0

File: .github/workflows/build-and-release-script.yml [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+      - name: Import GPG key
+        uses: crazy-max/ghaction-import-gpg@01dd5d3ca463c7f10f7f4f7b4f177225ac661ee4 # v6.1.0
+        id: import_gpg
+        with:
+          gpg_private_key: ${{ secrets.GPG_PRIVATE_KEY }}
+          passphrase: ${{ secrets.PASSPHRASE }}
+      - name: Run GoReleaser
+        uses: goreleaser/goreleaser-action@5742e2a039330cbb23ebf35f046f814d4c6ff811 # v5.1.0
+        with:
+          args: release --clean
```
Comment (vandyliu) [inline .github/workflows/build-and-release-script.yml:58]: how does it know which file to build :O

File: import/import.go [hunk 8]
```diff
@@ -0,0 +1,566 @@
...
+	Resource string
+	Error    error
+}
+
+func main() {
+	log.SetFlags(0)
+	log.Println("Terraform Import Script Starting")
+
+	// collect all arguments from the user, indicating all the resources that need to be imported
+	resourcesPtr := flag.String("resources", "", "Comma separated list of resources to import")
```
Comment (vandyliu) [inline import/import.go:33]: maybe we shuold have the list in the help part as well

File: .github/workflows/build-and-release-script.yml [hunk 9]
```diff
@@ -0,0 +1,67 @@
...
+  lint:
+    name: Lint
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@a5ac7e51b41094c92402da3b24376905380afc29 # v4.1.6
+      - uses: actions/setup-go@cdcb36043654635271a94b9a6d1392de5bb323a7 # v5.0.1
+        with:
+          go-version-file: 'go.mod'
+          cache: true
+      - run: make validate-fmt
```
Comment (vandyliu) [inline .github/workflows/build-and-release-script.yml:36]: we should make sure this is only for this file

File: .github/workflows/build-and-release-script.yml [hunk 10]
```diff
@@ -0,0 +1,67 @@
+name: Build and Release
+
+on:
+  push:
+    tags:
+      - 'v*'
```
Comment (vandyliu) [inline .github/workflows/build-and-release-script.yml:6]: hmm maybe we can have a different tag for this import

File: .github/workflows/build-and-release-script.yml (hunk 7 above)
Comment (ichung08) [reply .github/workflows/build-and-release-script.yml:58]: i think with this current configuration it builds everything, need to look into it more

File: .github/workflows/build-and-release-script.yml
Comment (vandyliu) [inline .github/workflows/build-and-release-script.yml:1]: build-release-import-script.yml
prob better name

File: import/import.go [hunk 11]
```diff
@@ -0,0 +1,566 @@
...
+	"github.com/astronomer/terraform-provider-astro/internal/clients/platform"
+	"github.com/samber/lo"
+)
+
+type HandlerResult struct {
+	Resource string
+	Error    error
+}
+
+func main() {
```
Comment (vandyliu) [inline import/import.go:28]: we should have some tests for this script
to assert the output file is correct

Comment (vandyliu) [reply import/import.go:28]: either mock the client calls or do real ones, im cool w either

File: Makefile [hunk 12]
```diff
@@ -20,12 +23,27 @@ test:
...
+
 .PHONY: fmt
 fmt:
 	gofmt -w ./
 	goimports -w ./
 	[ -z "$$CIRCLECI" ] || git diff --exit-code --color=always # In CI: exit if anything changed
 
+.PHONY: mock
+mock:
+	@if [[ -z "$MOCKERY_VERSION" || ! "$MOCKERY_VERSION" =~ .*v2\.42\.0.* ]]; then \
```
Comment (vandyliu) [inline Makefile:39]: can we format this properly

File: Makefile [hunk 13]
```diff
@@ -9,6 +9,9 @@ $(ENVTEST_ASSETS_DIR):
 	mkdir -p $(ENVTEST_ASSETS_DIR)
 OAPI_CODEGEN ?= $(ENVTEST_ASSETS_DIR)/oapi-codegen
 
+MOCKERY_VERSION := 2.34.0
+MOCKERY_DOWNLOAD_DIR := /tmp/mockery-$(MOCKERY_VERSION)
```
Comment (vandyliu) [inline Makefile:13]: we should just use `ENVTEST_ASSETS_DIR` instead of `MOCKERY_DOWNLOAD_DIR`

File: Makefile [hunk 14]
```diff
@@ -9,6 +9,9 @@ $(ENVTEST_ASSETS_DIR):
 	mkdir -p $(ENVTEST_ASSETS_DIR)
 OAPI_CODEGEN ?= $(ENVTEST_ASSETS_DIR)/oapi-codegen
 
+MOCKERY_VERSION := 2.34.0
```
Comment (vandyliu) [inline Makefile:12]: it says 2.34.0 here but in the make mock it says 2.42.0

File: Makefile [hunk 15]
```diff
@@ -20,12 +23,27 @@ test:
 	go vet ./...
 	TF_ACC="" go test ./... -v $(TESTARGS)
 
+# Run script tests
+.PHONY: testscript
+testscript:
```
Comment (vandyliu) [inline Makefile:28]:
```suggestion
.PHONY: test-import-script
test-import-script:
```

File: .github/workflows/import-script-tests.yml [hunk 16]
```diff
@@ -0,0 +1,27 @@
...
+    timeout-minutes: 5
+    steps:
+      - uses: actions/checkout@a5ac7e51b41094c92402da3b24376905380afc29 # v4.1.6
+      - uses: actions/setup-go@cdcb36043654635271a94b9a6d1392de5bb323a7 # v5.0.1
+        with:
+          go-version-file: 'go.mod'
+          cache: true
+      - run: go mod download
+      - run: go build -v .
+      - run: make testscript
```
Comment (vandyliu) [inline .github/workflows/import-script-tests.yml:27]:
```suggestion
      - run: make test-import-script
```

File: import/import_script_test.go [hunk 17]
```diff
@@ -0,0 +1,379 @@
...
+				},
+			}
+
+			mockPlatformClient.On("ListWorkspacesWithResponse", ctx, organizationId, (*platform.ListWorkspacesParams)(nil)).Return(mockResponse, nil)
+
+			result, err := import_script.HandleWorkspaces(ctx, mockPlatformClient, mockIAMClient, organizationId)
+
+			Expect(err).To(BeNil())
+			Expect(result).To(ContainSubstring(fmt.Sprintf("astro_workspace.workspace_%s", workspaceId1)))
+			Expect(result).To(ContainSubstring(fmt.Sprintf("astro_workspace.workspace_%s", workspaceId2)))
```
[the remark this answers is not in the document]
Comment (ichung08) [reply import/import_script_test.go:76]: no for those I think we would need like an integration test

these tests only test the output of each individual handler that generate import blocks

**Review by vandyliu — APPROVED** [review]

File: .github/workflows/build-release-import-script.yml [hunk 18]
```diff
@@ -0,0 +1,65 @@
+name: Build and Release Import Script
+
+on:
+  push:
+    tags:
+      - 'import-v*'
```
Comment (vandyliu) [inline .github/workflows/build-release-import-script.yml:6]: dont forget to test this after merge

File: README.md [hunk 19]
```diff
@@ -203,5 +203,59 @@ You will also need to set all the environment variables described in `internal/p
...
+The Astro Terraform Import Script is a tool designed to help you import existing Astro resources into your Terraform configuration. 
+Currently, this script automates the process of generating Terraform import blocks and resource configurations for the following resources: workspaces, deployments, clusters, hybrid cluster workspace authorizations, API tokens, teams, team roles, and user roles.
 
+To use the import script, run it with the following syntax:
+
+```
+go run ./import/import_script.go [options]
+```
+
+Additionally, you can build the script into a binary and run it as an executable:
```
Comment (vandyliu) [inline README.md:216]: there should be a downloadable executable created by the release GH action and that is prob the best way to use it

File: README.md [hunk 20]
```diff
@@ -203,5 +203,59 @@ You will also need to set all the environment variables described in `internal/p
...
+   ```
+
+2. Import all supported resources and run Terraform init:
+   ```
+   go run import_script.go -resources=workspace,deployment,cluster,api_token,team,team_roles,user_roles -token=your_api_token -organizationId=your_org_id -runTerraformInit
+   ```
+
+3. Use a different API host (e.g., dev environment):
+   ```
+   go run import_script.go -resources=workspace -token=your_api_token -organizationId=your_org_id -host=dev
```
Comment (vandyliu) [inline README.md:246]: ideally these would be an `exec` file like `./import -resources=workspace`

File: import/import_script.go [hunk 21]
```diff
@@ -0,0 +1,897 @@
...
+	}
+
+	if len(token) == 0 {
+		log.Fatal("API token not provided")
+		return
+	}
+
+	// set the host
+	var host string
+	if *hostPtr == "dev" {
```
Comment (vandyliu) [inline import/import_script.go:76]: this does not match readme nor line 38

File: import/import_script.go [hunk 22]
```diff
@@ -0,0 +1,897 @@
...
+			if err != nil {
+				log.Printf("Error handling resource %s: %v", resource, err)
+				results <- HandlerResult{Resource: resource, Error: err}
+			} else {
+				results <- HandlerResult{Resource: resource, ImportString: result}
+			}
+		}(resource)
+	}
+
+	go func() {
```
Comment (vandyliu) [inline import/import_script.go:164]: curious why we need a goroutine here

File: import/import_script.go [hunk 23]
```diff
@@ -0,0 +1,897 @@
...
+	// write the terraform configuration to a file
+	err = os.WriteFile("import.tf", []byte(importString), 0644)
+	if err != nil {
+		log.Fatalf("Failed to write import configuration to file: %v", err)
+		return
+	}
+
+	log.Println("Successfully wrote import configuration to import.tf")
+
+	// Trigger terraform init if the flag is set
```
Comment (vandyliu) [inline import/import_script.go:194]: we should comment why we added this
ie. for our integration tests in ci

File: import/import_script.go (hunk 21 above)
Comment (vandyliu) [reply import/import_script.go:76]: actually its fine nvm

File: import/import_script.go [hunk 24]
```diff
@@ -0,0 +1,897 @@
...
+	for clusterId, clusterType := range clusterMap {
+		if clusterType != platform.ClusterTypeHYBRID {
+			clusterImportString = fmt.Sprintf(`
+import {
+	id = "%v"
+	to = astro_cluster.cluster_%v
+}`, clusterId, clusterId)
+		}
+
+		if clusterType == platform.ClusterTypeHYBRID {
```
Comment (vandyliu) [inline import/import_script.go:440]: nit we can just use else here

File: import/import_script.go [hunk 25]
```diff
@@ -0,0 +1,897 @@
...
+}`, apiTokenId, apiTokenId)
+
+		importString += apiTokenImportString + "\n"
+	}
+
+	return importString, nil
+}
+
+func handleTeams(ctx context.Context, platformClient *platform.ClientWithResponses, iamClient *iam.ClientWithResponses, organizationId string) (string, error) {
+	log.Printf("Importing teams for organization %s", organizationId)
```
Comment (vandyliu) [inline import/import_script.go:501]: do we need to worry about scim for these

File: import/import_script.go [hunk 26]
```diff
@@ -0,0 +1,897 @@
...
+	err = os.WriteFile("generated.tf", []byte(newContent), 0644)
+	if err != nil {
+		return fmt.Errorf("failed to write updated generated.tf: %v", err)
+	}
+
+	log.Println("Successfully updated generated.tf with deployment information.")
+	return nil
+}
+
+func generateDeploymentHCL(ctx context.Context, platformClient *platform.ClientWithResponses, organizationId string) (string, error) {
```
Comment (vandyliu) [inline import/import_script.go:679]: lets add comment for why we need this

Comment (vandyliu) [inline import/import_script.go:679]: whats HCL mean here

File: import/import_script_mock_functions.go [hunk 27]
```diff
@@ -0,0 +1,349 @@
...
+	"github.com/astronomer/terraform-provider-astro/internal/clients"
+	"github.com/astronomer/terraform-provider-astro/internal/clients/iam"
+	"github.com/astronomer/terraform-provider-astro/internal/clients/platform"
+	mocksIam "github.com/astronomer/terraform-provider-astro/internal/mocks/iam"
+	mocksPlatform "github.com/astronomer/terraform-provider-astro/internal/mocks/platform"
+	"github.com/samber/lo"
+	"golang.org/x/exp/maps"
+)
+
+func HandleWorkspaces(ctx context.Context, platformClient *mocksPlatform.ClientWithResponsesInterface, iamClient *mocksIam.ClientWithResponsesInterface, organizationId string) (string, error) {
```
Comment (vandyliu) [inline import/import_script_mock_functions.go:18]: can all these functions just go in the test file instead?

File: import/import_script.go (hunk 25 above)
Comment (ichung08) [reply import/import_script.go:501]: yes we should exit if scim

File: import/import_script.go (hunk 26 above)
Comment (ichung08) [reply import/import_script.go:679]: HashiCorp Configuration Language

File: import/import_script_mock_functions.go (hunk 27 above)
Comment (ichung08) [reply import/import_script_mock_functions.go:18]: yes, i wanted to separate it so its cleaner but i can move them all in there too

File: import/import_script.go (hunk 22 above)
Comment (ichung08) [reply import/import_script.go:164]: the block above parallelizes the import generation step for each resource in individual go routines

to ensure that all the go routines are finished, we need a separate goroutine to wait for all the goroutines to finish writing to the channel before closing the channel

File: import/import_script.go (hunk 21 above)
Comment (ichung08) [reply import/import_script.go:76]: what about it doesn't match?

File: import/import_script.go (hunk 22 above)
Comment (vandyliu) [reply import/import_script.go:164]: would it work the same without the goroutine?
to me, it seems like without go routine makes more sense so just 
```
wg.Wait()
close(results)
```

but maybe im missing something

**Review by vandyliu — APPROVED** [review]

**Pull request merged**